File: 백준/Gold/3055. 탈출/탈출.py
from collections import deque

R,C = map(int,input().split())
forest = [list(input()) for _ in range(R)]
visited = [[0]*C for _ in range(R)]

#상하좌우
dx = [-1,1,0,0]
dy = [0,0,-1,1]
q = deque()

def bfs(a,b):
    while q:
        x,y = q.popleft()
        if forest[a][b] == "S":
            return visited[a][b]
        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]
            if 0<=nx<R and 0<=ny<C:
                if forest[nx][ny] == "." or forest[nx][ny] == 'D':
                    if forest[x][y]=="S":
                        forest[nx][ny] = "S"
                        q.append((nx,ny))
                        visited[nx][ny] = visited[x][y] + 1
                if forest[nx][ny] == "." or forest[nx][ny] == 'S':
                    if forest[x][y] == "*":
                        q.append((nx,ny))
                        forest[nx][ny] = "*"
                
    return "KAKTUS"       
                 

for i in range(R):
    for j in range(C):
        if forest[i][j] == 'S':
            q.append((i,j))
        elif forest[i][j] == "D":
            ax,ay = i,j
            
for x in range(R):
    for y in range(C):
        if forest[x][y] == "*":
            q.append((x,y))


print(bfs(ax,ay))    

File: 백준/Gold/3055. 탈출/탈출.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static int R,C;
	static int gx,gy;
	static char[][] board;
	static Queue<Integer[]> q;
	static int[][] visited;
	static int[] dx = {-1,1,0,0};
	static int[] dy = {0,0,-1,1};
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		R = Integer.parseInt(st.nextToken());
		C = Integer.parseInt(st.nextToken());
		q = new LinkedList<>();
		visited = new int[R][C];
		board = new char[R][C];
		
		for(int i = 0 ; i < R ; i++) {
			String a = br.readLine();
			for(int j = 0 ; j < C ; j++) {
				board[i][j] = a.charAt(j);
			}
		}
		
		for(int i = 0 ; i < R; i++) {
			for(int j = 0 ; j < C; j++) {
				if(board[i][j] == 'S') {
					q.add(new Integer[] {i,j});
				}else if(board[i][j] == 'D') {
					gx = i;
					gy = j;
				}
			}
		}
		
		for(int i = 0 ; i < R; i++) {
			for(int j = 0; j < C; j++) {
				if(board[i][j] == '*') {
					q.add(new Integer[] {i,j});
				}
			}
		}
		
		
		int ans = bfs(gx,gy);
		if(ans == -1) {
			System.out.println("KAKTUS");
		}else {
			System.out.println(ans);
		}

	}
	private static int bfs(int gx, int gy) {
		while(!q.isEmpty()) {
			int x = q.peek()[0];
			int y = q.peek()[1];
			q.poll();
			if(board[gx][gy] == 'S') {
				return visited[gx][gy];
			}
			
			for(int i = 0 ; i < 4; i++) {
				int nx = x + dx[i];
				int ny = y + dy[i];
				if(0 <= nx && nx < R && 0 <= ny && ny < C) {
					if(board[nx][ny] == '.' || board[nx][ny] == 'D') {
						if(board[x][y] == 'S') {
							board[nx][ny] = 'S';
							visited[nx][ny] = visited[x][y] + 1;
							q.add(new Integer[] {nx,ny});
						}
					}
					if(board[nx][ny] == '.' || board[nx][ny] == 'S') {
						if(board[x][y] == '*') {
							board[nx][ny] = '*';
							q.add(new Integer[] {nx,ny});
						}
					}
					
				}
			}

		}
		
		return -1;
		
	}

}